docs/ci.md

```diff
@@ -26,24 +33,125 @@ steps:
+language: node_js
+dist: bionic
+addons:
+  apt:
+    packages:
+    # These are required to run webkit
+    - libwoff1
+    - libopus0
+    - libwebp6
+    - libwebpdemux2
+    - libenchant1c2a
+    - libgudev-1.0-0
+    - libsecret-1-0
+    - libhyphen0
+    - libgdk-pixbuf2.0-0
+    - libegl1
+    - libgles2
+    - libevent-2.1-6
+    - libnotify4
+    - libxslt1.1
+    - libvpx5
+    # This is required to run chromium
+    - libgbm1
+
+# allow headful tests
+before_install:
+  # Enable user namespace cloning
+  - "sysctl kernel.unprivileged_userns_clone=1"
+  # Launch XVFB
+  - "export DISPLAY=:99.0"
+  - "sh -e /etc/init.d/xvfb start"
+```
+
+### CircleCI
+
+We run our tests on CircleCI, with our [pre-built Docker image](docker/README.md). Use our [CircleCI configuration](/.circleci/config.yml) to create your own. Running Playwright smoothly on CircleCI requires the following steps:
+
+1. Use the pre-built [Docker image](docker/README.md) in your config like so:
+
+   ```yaml
+   docker:
+     - image: aslushnikov/playwright:bionic
+       environment:
+         NODE_ENV: development # Needed if playwright is in `devDependencies`
+   ```
+
+1. If you’re using Playwright through Jest, then you may encounter an error spawning child processes:
+
+   ```
+   [00:00.0]  jest args: --e2e --spec --max-workers=36
+   Error: spawn ENOMEM
+      at ChildProcess.spawn (internal/child_process.js:394:11)
+   ```
+
+   This is likely caused by Jest autodetecting the number of processes on the entire machine (`36`) rather than the number allowed to your container (`2`). To fix this, set `jest --maxWorkers=2` in your test command.
+
+### AppVeyor
+
+We run our tests on Windows agents in AppVeyor. Use our [AppVeyor configuration](/.appveyor.yml) to create your own.
+
+## Debugging browser launches
+
+Playwright supports the `DEBUG` environment variable to output debug logs during execution. Setting it to `pw:browser*` is helpful while debugging `Error: Failed to launch browser` errors.
+
+```
+DEBUG=pw:browser* npm run test
+```
+
+## Caching browsers
 
-We run our tests on CircleCI, with our [pre-built Docker image](docker/README.md). Use our [CircleCI configuration](/.circleci/config.yml) to create your own.
+By default, Playwright installs browser binaries in the following directories. This behavior can be [customized with environment variables](installation.md).
 
-## AppVeyor
+- `%USERPROFILE%\AppData\Local\ms-playwright` on Windows
+- `~/Library/Caches/ms-playwright` on MacOS
+- `~/.cache/ms-playwright` on Linux
 
-We run our tests on Windows agents in AppVeyor. Use our [AppVeyor configuration](/.appveyor.yml) to create your own.
\ No newline at end of file
+These locations are not covered by typical CI configurations, which cache the project `node_modules` or the [npm-cache directory](https://docs.npmjs.com/cli-commands/cache.html). To cache the browser binaries between CI runs, these locations can be explicitly cached in your CI configuration.
```

Comment:
We should mention that the hash should include the Playwright version to make sure they don't wipe each other